### PR TITLE
grpc enabling

### PR DESCRIPTION
**Pull request by szc321**
Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/lava/magma/runtime/message_infrastructure/test/test_channel.py [hunk 1]
```diff
@@ -172,6 +175,49 @@ def test_single_process_socketchannel(self):
...
+    def test_single_process_socketchannel(self):
+
+        size = 5
+        predata = prepare_data()
+        predata2 = prepare_data()
+        print(predata)
+        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
+        name = 'test_single_process_socket_channel'
+
+        socket_channel = Channel(
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/test/test_channel.py:187]: Create new ut for grpc channel. Do not reuse socketchannel.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  std::string AllocChannelGrpc(size_t nbytes) {
+    std::string url = base_url + std::to_string(url_num) \
+    + base_port+std::to_string(port_num);
+    urls_.push_back(url);
+    url_num++;
+    port_num++;
+    return url;
+  }
+
+friend GrpcManager &GetGrpcManager();
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h:39]: use space ahead.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+namespace message_infrastructure {
+
+class GrpcManager {
+ public:
+  ~GrpcManager();
+
+  std::string AllocChannelGrpc(size_t nbytes) {
+    std::string url = base_url + std::to_string(url_num) \
+    + base_port+std::to_string(port_num);
+    urls_.push_back(url);
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h:33]: code style. use space before and after + ......

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  //}
+}
+bool GrpcSendPort::Probe() {
+  return false;
+}
+void GrpcSendPort::Join() {
+  done_ = true;
+}
+
+}  // namespace message_infrastructure
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:229]: blank space at last line.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+#include <message_infrastructure/csrc/core/message_infrastructure_logging.h>
+#include <grpcpp/ext/proto_server_reflection_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#include <memory>
+#include <iostream>
+#include <string>
+
+#include "message_infrastructure/csrc/channel/grpc/build/grpcchannel.grpc.pb.h"
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:16]: Use same style for header files.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/socket/socket_channel.cc [hunk 6]
```diff
@@ -12,6 +12,7 @@ namespace message_infrastructure {
 SocketChannel::SocketChannel(const std::string &src_name,
                              const std::string &dst_name,
                              const size_t &nbytes) {
+
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/socket/socket_channel.cc:15]: rm useless blank line.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#include <message_infrastructure/csrc/core/message_infrastructure_logging.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <memory>
+#include <set>
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h:18]: rm useless headers.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+using grpcchannel::GrpcChannelServer;
+using grpcchannel::GrpcMetaData;
+using GrpcMetaDataPtr = std::shared_ptr<GrpcMetaData>;
+
+GrpcChannelServerImpl::GrpcChannelServerImpl(const std::string& name,
+                                             const size_t &size,
+                                             const size_t &nbytes)
+                        :name_(name), size_(size),
+                        nbytes_(nbytes), done_(false) {
+  recvqueue = std::make_shared<RecvQueue<GrpcMetaDataPtr>>(name_, size_, nbytes_); //NOLINT
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:36]: Delete //nolint comment.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+void GrpcChannelServerImpl::Stop() {
+  done_ = true;
+  recvqueue->Stop();
+}
+
+
+GrpcRecvPort::GrpcRecvPort(const std::string& name,
+                 const size_t &size,
+                 const size_t &nbytes, const std::string& url)\
+                 :AbstractRecvPort(name, size, nbytes), done_(false), url_(url) { //NOLINT
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:71]: Delete //NOLINT comment.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  channel = grpc::CreateChannel(url_, grpc::InsecureChannelCredentials());
+  stub_ = GrpcChannelServer::NewStub(channel);
+}
+GrpcMetaData GrpcSendPort::MetaData2GrpcMetaData(MetaDataPtr metadata) {
+  GrpcMetaData grpcdata;
+  grpcdata.set_nd(metadata->nd);
+  grpcdata.set_type(metadata->type);
+  grpcdata.set_elsize(metadata->elsize);
+  grpcdata.set_total_size(metadata->total_size);
+  char* data = (char*)metadata->mdata; //NOLINT
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:135]: Delete NOLINT comment and fix lint error.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+void GrpcSendPort::Send(MetaDataPtr metadata) {
+  GrpcMetaData request = MetaData2GrpcMetaData(metadata);
+  DataReply reply;
+  ClientContext context;
+  Status status = stub_->RecvArrayData(&context, request, &reply);
+  // if (status.ok()&&reply.ack()) {
+  //    std::cout << "Successed, Recv ack is " << reply.ack() << std::endl;
+  //} else {
+  //    std::cout << "ERROR! Send fail!" << std::endl;
+  //}
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:152]: Remove useless code.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/shmem/shmem_port.h [hunk 12]
```diff
@@ -37,7 +38,7 @@ class ShmemSendPort final : public AbstractSendPort {
 };
 
 using ShmemSendPortPtr = std::shared_ptr<ShmemSendPort>;
-
+/*
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/shmem/shmem_port.h:41]: Remove useless code

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/common.h [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+#include <memory>
+#include <string>
+#include <cassert>
+
+namespace message_infrastructure {
+
+template<class T>
+class RecvQueue{
+ public:
+  explicit RecvQueue(const std::string& name,
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/common.h:21]: Constructor functions with more than 1 paramters do not need to use 'explicit'.

File: src/lava/magma/runtime/message_infrastructure/test/test_channel.py [hunk 14]
```diff
@@ -164,14 +166,71 @@ def test_single_process_socketchannel(self):
...
+        resdata = recv_port.recv()
+
+        if not np.array_equal(resdata, predata):
+            raise AssertionError()
+   
+        send_port.join()
+        recv_port.join()
 if __name__ == "__main__":
-    unittest.main()
+    unittest.main()
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/test/test_channel.py:236]: use $flakeheaven lint src/lava tests/ to check python file lint errors.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.h [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+  void Start();
+  void Send(MetaDataPtr metadata);
+  void Join();
+  bool Probe();
+  GrpcMetaData MetaData2GrpcMetaData(MetaDataPtr metadata);
+
+ private:
+  std::shared_ptr<Channel> channel;
+  std::atomic_bool done_;
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.h:100]: useless value?

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_channel.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace message_infrastructure {
+
+class GrpcChannel : public AbstractChannel {
+ public:
+  GrpcChannel() {}
+  GrpcChannel(const std::string &src_name,
+              const std::string &dst_name,
+              const size_t &size,
+              const size_t &nbytes);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_channel.h:22]: remove nbytes

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+bool GrpcRecvPort::Probe() {
+  return serviceptr->Probe();
+}
+void GrpcRecvPort::GrpcMetaData2MetaData(MetaDataPtr metadata,
+                                         GrpcMetaDataPtr grpcdata) {
+  metadata->nd = grpcdata->nd();
+  metadata->type = grpcdata->type();
+  metadata->elsize = grpcdata->elsize();
+  metadata->total_size = grpcdata->total_size();
+  void* data = malloc(nbytes_);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:118]: calculate nbytes_ by elsize and type...

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+  grpcdata.set_nd(metadata->nd);
+  grpcdata.set_type(metadata->type);
+  grpcdata.set_elsize(metadata->elsize);
+  grpcdata.set_total_size(metadata->total_size);
+  char* data = reinterpret_cast<char*>(metadata->mdata);
+  for (int i = 0; i < MAX_ARRAY_DIMS; i++) {
+    grpcdata.add_dims(metadata->dims[i]);
+    grpcdata.add_strides(metadata->strides[i]);
+  }
+  grpcdata.set_value(data, nbytes_);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.cc:146]: same with `GrpcRecvPort::GrpcMetaData2MetaData`

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_channel.cc [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+GrpcChannel::GrpcChannel(const std::string &url,
+                         const int &port,
+                         const std::string &src_name,
+                         const std::string &dst_name,
+                         const size_t &size) {
+  std::string url_ = url + std::to_string(port);
+  bool ret = GetGrpcManager().CheckURL(url_);
+  if (!ret) {
+    // maybe throw an exception?
+    std::cout << "url is used" << std::endl;
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_channel.cc:23]: Print a log.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.h [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+               url_(url) {}
+
+  void Start();
+  void Send(MetaDataPtr metadata);
+  void Join();
+  bool Probe();
+  GrpcMetaData MetaData2GrpcMetaData(MetaDataPtr metadata);
+
+ private:
+  std::shared_ptr<Channel> channel;
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc_port.h:101]: Code style. `channel_`.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/ports.py [hunk 21]
```diff
@@ -4,6 +4,7 @@
 
 from MessageInfrastructurePywrapper import SendPort as CppSendPort
 from MessageInfrastructurePywrapper import Channel as CppChannel
+from MessageInfrastructurePywrapper import GetRPCChannel as CppRPCChannel
```
[the remark this answers is not in the document]
Comment (szc321) [reply src/lava/magma/runtime/message_infrastructure/message_infrastructure/ports.py:7]: If I delete this, there is a  ImportError: cannot import name 'GetRPCChannel' from 'message_infrastructure.ports' (/home/szc/LAVA/lava_nc_grpc_szc/lava/src/lava/magma/runtime/message_infrastructure/message_infrastructure/ports.py)

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/channel_factory.h [hunk 22]
```diff
@@ -37,8 +37,23 @@ class ChannelFactory {
...
+
   friend ChannelFactory& GetChannelFactory();
 
+  AbstractChannelPtr GetRPCChannel(const std::string &url,
+                                   const int &port,
+                                   const std::string &src_name,
+                                   const std::string &dst_name,
+                                   const size_t &size) {
+  return std::make_shared<GrpcChannel>(url, port, src_name, dst_name, size);
+}
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/channel_factory.h:49]: Position of `}`

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/channel_factory.h [hunk 23]
```diff
@@ -37,8 +37,23 @@ class ChannelFactory {
...
+
+  AbstractChannelPtr GetDefRPCChannel(const std::string &src_name,
+                                      const std::string &dst_name,
+                                      const size_t &size) {
+  return std::make_shared<GrpcChannel>(src_name, dst_name, size);
+}
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/channel_factory.h:55]: Position of `}`

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    if (url_set_.count(url))
+      return false;
+    url_set_.insert(url);
+    return true;
+  }
+
+  std::string AllocURL() {
+    std::string url = base_url_ + std::to_string(url_num_) \
+                      + base_port_+std::to_string(port_num_);
+    urls_.push_back(url);
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h:31]: use 1 container to store the urls for both AllocURL and CheckURL.

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+ public:
+  ~GrpcManager();
+  bool CheckURL(const std::string &url) {
+    if (url_set_.count(url))
+      return false;
+    url_set_.insert(url);
+    return true;
+  }
+
+  std::string AllocURL() {
```
Comment (hexu33) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/channel/grpc/grpc.h:28]: `AllocURL` should also check url first.

**Pull request merged**